File: backend/services/screener_engine/compiler.js
const fieldMap = require("./field_mapping");

function compileCondition(condition, params) {
  const column = fieldMap[condition.field];
  if (!column) throw new Error("Invalid field");

  params.push(condition.value);

  switch (condition.operator) {
    case "<": return `${column} < $${params.length}`;
    case ">": return `${column} > $${params.length}`;
    case "=": return `${column} = $${params.length}`;
    default:
      throw new Error("Unsupported operator");
  }
}

function compileFilter(filter, params) {
  if (filter.and) {
    return filter.and
      .map(c => compileCondition(c, params))
      .join(" AND ");
  }

  if (filter.or) {
    return filter.or
      .map(c => compileCondition(c, params))
      .join(" OR ");
  }

  throw new Error("Invalid filter");
}

module.exports = { compileFilter };

File: backend/screener_engine/compiler.py
ALLOWED_FIELDS = {
    "pe_ratio",
    "promoter_holding",
    "net_profit",
    "revenue",
    "close_price"
}

ALLOWED_OPERATORS = {
    ">", "<", ">=", "<=", "=", "!="
}

def compile_condition(condition, params):
    field = condition.get("field")
    operator = condition.get("operator")
    value = condition.get("value")

    if field not in ALLOWED_FIELDS:
        raise ValueError(f"Unsupported field: {field}")

    if operator not in ALLOWED_OPERATORS:
        raise ValueError(f"Unsupported operator: {operator}")

    sql_op = operator
    params.append(value)

    return f"{field} {sql_op} %s"


def compile_filter(filter_block, params):
    if "and" in filter_block:
        clauses = [
            compile_filter(c, params) for c in filter_block["and"]
        ]
        return "(" + " AND ".join(clauses) + ")"

    if "or" in filter_block:
        clauses = [
            compile_filter(c, params) for c in filter_block["or"]
        ]
        return "(" + " OR ".join(clauses) + ")"

    return compile_condition(filter_block, params)


def compile_dsl(dsl):
    params = []
    where_clause = compile_filter(dsl["filter"], params)
    return where_clause, params
